batch_requests.go

```diff
@@ -0,0 +1,210 @@
+				if err != nil {
+					return res, err
+				}
+				err = json.Unmarshal(jsonStr, &errResp)
+				if err != nil {
+					return res, err
+				}
+				return res, errResp.Error
+			}
+
+			jsonStr, err := json.Marshal(resp.Body)
+			if err != nil {
+				return res, err
+			}
+			err = json.Unmarshal(jsonStr, &res)
+			if err != nil {
+				return res, err
+			}
+
+			return res, nil
+		}
+	}
+
+	return res, errors.New("Response not found, check if id is valid")
+}
+
+// NewBatchRequest creates an instance of BatchRequest
+func NewBatchRequest() *batchRequest {
+	return &batchRequest{}
+}
+
+func toBatchItem(requestInfo abstractions.RequestInformation) (*batchItem, error) {
+	uri, err := requestInfo.GetUri()
+	if err != nil {
+		return nil, err
+	}
+
+	var body map[string]any
+	err = json.Unmarshal(requestInfo.Content, &body)
+	if err != nil {
+		return nil, err
+	}
+
+	return &batchItem{
+		Id:        uuid.NewString(),
+		Method:    requestInfo.Method.String(),
+		Body:      body,
+		Headers:   requestInfo.Headers,
+		Url:       uri.Path,
+		DependsOn: make([]string, 0),
+	}, nil
+}
+
+func (br *batchRequest) toJson() ([]byte, error) {
+	return json.Marshal(br)
+}
+
+func getBaseUrl(adapter abstractions.RequestAdapter) (*url.URL, error) {
+	return url.Parse(adapter.GetBaseUrl())
+}
+
+func buildRequestInfo(jsonBody []byte, baseUrl *url.URL) *abstractions.RequestInformation {
+	requestInfo := abstractions.NewRequestInformation()
+
+	requestInfo.SetStreamContent(jsonBody)
+	requestInfo.Method = abstractions.POST
+	requestInfo.SetUri(*baseUrl)
+	requestInfo.Headers = map[string]string{
+		"Content-Type": "application/json",
+	}
+
+	return requestInfo
+}
+
+func sendBatchRequest(requestInfo *abstractions.RequestInformation, adapter abstractions.RequestAdapter) (*BatchResponse, error) {
+	if requestInfo == nil {
+		return nil, errors.New("requestInfo cannot be nil")
+	}
+
+	// SendAsync type asserts HandlerFunc's return value into a parsable. We bypass SendAsync deserialization by returning a noop
+	// parsable struct and directly marshalling the response into a struct.
+	var res BatchResponse
+	_, err := adapter.SendAsync(requestInfo, nil, func(response any, errorMappings abstractions.ErrorMappings) (any, error) {
+		resp, ok := response.(*http.Response)
+		if !ok {
+			return nil, errors.New("Response type assertion failed")
+		}
+
+		if status := resp.StatusCode; status >= 400 && status < 600 {
+			return nil, fmt.Errorf("Request failed with status: %d", status)
+		}
+
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		err = json.Unmarshal(body, &res)
+		if err != nil {
+			return nil, err
+		}
+
+		return internal.NewNoOpParsable(), nil
+	}, nil)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &res, nil
+}
+
+type batchRequest struct {
+	Requests []*batchItem `json:"requests"`
+}
+
+type BatchResponse struct {
+	Responses []batchItem
+}
+
+type errorDetails struct {
```

Comment:
The errors will be standard odata errors like we generate in the service SDK. Now the question becomes how do we perform the mapping?

Reply:
Since error mapping code is generated at the service SDK level, we can have a parameter to pass a custom error mapper by users and leave the errors to be cascaded without a default mapper. 

i.e using the method from core directly will throw an error since no mapper is provided.

At core lib level , we only provide

```go
batch.Send(reqAdapter, errorMapper)
```

At the service library level a default error mapper will be introduced, this should allow the format 

```go
graphClient.SendBatch(batch)
```

or

```go
graphClient.SendBatchWithMapper(batch, errorMapper)
```

Reply:
This avenue probably makes using the batch requests a bit more complex for end users than it needs to be.
An alternative would be to have a singleton map in core like `BatchingErrorMapping := make(map[string]ParsableFactory)`

In the service lib we'd have the handcrafted graph request adapter register the errors for the status codes onto that singleton in the constructor.

This way no additional work for end users, and they need the request adapter to make batch requests anyway. What do you think?

Reply:
Made the changes to support this feature